File: backend/services/stigmer-server/pkg/domain/workflowexecution/controller/get_execution_summary.go
package workflowexecution

import (
	"context"
	"time"

	workflowexecutionv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/workflowexecution/v1"
	"github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/commons/apiresource/apiresourcekind"
	grpclib "github.com/stigmer/stigmer/backend/libs/go/grpc"
	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/durationpb"
)

// GetExecutionSummary returns aggregated execution statistics for the requested
// organization and time window. For OSS (single-user local), org filtering is
// a no-op — all executions are aggregated.
//
// @since T14 (Dashboard Integration)
func (c *WorkflowExecutionController) GetExecutionSummary(
	ctx context.Context,
	req *workflowexecutionv1.GetExecutionSummaryRequest,
) (*workflowexecutionv1.ExecutionSummary, error) {
	data, err := c.store.ListResources(ctx, apiresourcekind.ApiResourceKind_workflow_execution)
	if err != nil {
		return nil, grpclib.InternalError(err, "failed to list workflow executions for summary")
	}

	cutoff := resolveTimeCutoff(req.GetTimeWindow())

	phaseCounts := make(map[int32]int32)
	var activeCount int32
	var completedDurations []time.Duration
	failureCounts := make(map[string]int32)    // workflow slug → failure count
	workflowNames := make(map[string]string)   // workflow slug → name
	execByWorkflow := make(map[string]int32)   // workflow slug → execution count
	costByWorkflow := make(map[string]float64) // workflow slug → total cost USD

	var totalCostMicros int64
	var totalInputTokens int64
	var totalOutputTokens int64

	for _, d := range data {
		exec := &workflowexecutionv1.WorkflowExecution{}
		if err := proto.Unmarshal(d, exec); err != nil {
			continue
		}

		createdAt := auditCreatedAt(exec)
		if !createdAt.IsZero() && !cutoff.IsZero() && createdAt.Before(cutoff) {
			continue
		}

		phase := exec.GetStatus().GetPhase()
		phaseCounts[int32(phase)]++

		if isActivePhase(phase) {
			activeCount++
		}

		slug := extractWorkflowSlug(exec)
		if slug != "" {
			execByWorkflow[slug]++
			if name := exec.GetMetadata().GetName(); name != "" {
				workflowNames[slug] = name
			}
		}

		execCost := exec.GetStatus().GetTotalCostMicros()
		totalCostMicros += execCost
		totalInputTokens += exec.GetStatus().GetTotalInputTokens()
		totalOutputTokens += exec.GetStatus().GetTotalOutputTokens()
		if slug != "" && execCost > 0 {
			costByWorkflow[slug] += float64(execCost) / 1_000_000.0
		}

		if phase == workflowexecutionv1.ExecutionPhase_EXECUTION_COMPLETED {
			if d := completionDuration(exec); d > 0 {
				completedDurations = append(completedDurations, d)
			}
		}

		if phase == workflowexecutionv1.ExecutionPhase_EXECUTION_FAILED && slug != "" {
			failureCounts[slug]++
		}
	}

	summary := &workflowexecutionv1.ExecutionSummary{
		ActiveCount: activeCount,
		PhaseCounts: phaseCounts,
		TotalCost: &workflowexecutionv1.WorkflowCostSummary{
			TotalCostUsd:      float64(totalCostMicros) / 1_000_000.0,
			TotalInputTokens:  totalInputTokens,
			TotalOutputTokens: totalOutputTokens,
		},
	}

	if len(completedDurations) > 0 {
		var total time.Duration
		for _, d := range completedDurations {
			total += d
		}
		avg := total / time.Duration(len(completedDurations))
		summary.AvgDuration = durationpb.New(avg)
	}

	summary.TopFailingWorkflows = buildFailureRanks(failureCounts, workflowNames, 10)
	summary.CostByWorkflow = buildCostBreakdown(costByWorkflow, execByWorkflow, workflowNames, 10)

	return summary, nil
}

func resolveTimeCutoff(tw workflowexecutionv1.SummaryTimeWindow) time.Time {
	now := time.Now()
	switch tw {
	case workflowexecutionv1.SummaryTimeWindow_SUMMARY_TIME_WINDOW_LAST_24H:
		return now.Add(-24 * time.Hour)
	case workflowexecutionv1.SummaryTimeWindow_SUMMARY_TIME_WINDOW_LAST_7D:
		return now.Add(-7 * 24 * time.Hour)
	case workflowexecutionv1.SummaryTimeWindow_SUMMARY_TIME_WINDOW_LAST_30D:
		return now.Add(-30 * 24 * time.Hour)
	case workflowexecutionv1.SummaryTimeWindow_SUMMARY_TIME_WINDOW_ALL_TIME:
		return time.Time{}
	default:
		return now.Add(-7 * 24 * time.Hour)
	}
}

func isActivePhase(p workflowexecutionv1.ExecutionPhase) bool {
	return p == workflowexecutionv1.ExecutionPhase_EXECUTION_PENDING ||
		p == workflowexecutionv1.ExecutionPhase_EXECUTION_IN_PROGRESS ||
		p == workflowexecutionv1.ExecutionPhase_EXECUTION_PAUSED
}

func auditCreatedAt(exec *workflowexecutionv1.WorkflowExecution) time.Time {
	ts := exec.GetStatus().GetAudit().GetSpecAudit().GetCreatedAt()
	if ts == nil {
		return time.Time{}
	}
	return ts.AsTime()
}

func completionDuration(exec *workflowexecutionv1.WorkflowExecution) time.Duration {
	started := exec.GetStatus().GetStartedAt()
	completed := exec.GetStatus().GetCompletedAt()
	if started == "" || completed == "" {
		return 0
	}
	s, err1 := time.Parse(time.RFC3339, started)
	c, err2 := time.Parse(time.RFC3339, completed)
	if err1 != nil || err2 != nil {
		return 0
	}
	return c.Sub(s)
}

func extractWorkflowSlug(exec *workflowexecutionv1.WorkflowExecution) string {
	return exec.GetMetadata().GetSlug()
}

func buildFailureRanks(
	counts map[string]int32,
	names map[string]string,
	limit int,
) []*workflowexecutionv1.WorkflowFailureRank {
	type entry struct {
		slug  string
		count int32
	}
	entries := make([]entry, 0, len(counts))
	for slug, count := range counts {
		entries = append(entries, entry{slug, count})
	}
	sortDescInt32(entries, func(e entry) int32 { return e.count })
	if len(entries) > limit {
		entries = entries[:limit]
	}
	result := make([]*workflowexecutionv1.WorkflowFailureRank, len(entries))
	for i, e := range entries {
		result[i] = &workflowexecutionv1.WorkflowFailureRank{
			WorkflowSlug: e.slug,
			WorkflowName: names[e.slug],
			FailureCount: e.count,
		}
	}
	return result
}

func buildCostBreakdown(
	costs map[string]float64,
	execCounts map[string]int32,
	names map[string]string,
	limit int,
) []*workflowexecutionv1.WorkflowCostBreakdown {
	type entry struct {
		slug string
		cost float64
	}
	entries := make([]entry, 0, len(costs))
	for slug, cost := range costs {
		entries = append(entries, entry{slug, cost})
	}
	sortDescFloat64(entries, func(e entry) float64 { return e.cost })
	if len(entries) > limit {
		entries = entries[:limit]
	}
	result := make([]*workflowexecutionv1.WorkflowCostBreakdown, len(entries))
	for i, e := range entries {
		result[i] = &workflowexecutionv1.WorkflowCostBreakdown{
			WorkflowSlug:   e.slug,
			WorkflowName:   names[e.slug],
			TotalCostUsd:   e.cost,
			ExecutionCount: execCounts[e.slug],
		}
	}
	return result
}

func sortDescInt32[T any](s []T, key func(T) int32) {
	for i := 1; i < len(s); i++ {
		for j := i; j > 0 && key(s[j]) > key(s[j-1]); j-- {
			s[j], s[j-1] = s[j-1], s[j]
		}
	}
}

func sortDescFloat64[T any](s []T, key func(T) float64) {
	for i := 1; i < len(s); i++ {
		for j := i; j > 0 && key(s[j]) > key(s[j-1]); j-- {
			s[j], s[j-1] = s[j-1], s[j]
		}
	}
}

File: backend/services/stigmer-server/src/domain/workflowexecution/get-execution-summary.ts
/**
 * GetExecutionSummary — ports get_execution_summary.go (T14 dashboard,
 * T12 overview): a full scan aggregated into phase counts, cost totals,
 * average completed duration, top failure ranks, and per-workflow cost
 * breakdown, optionally scoped by time window and workflow.
 *
 * Zero-shape pins (Class A suite): success_rate is the -1 sentinel when
 * no terminal executions exist (distinguishable from a real 0%); the
 * zero-valued total_cost summary is ALWAYS present; avg_duration is
 * deliberately absent when nothing completed (an average over nothing is
 * not 0).
 *
 * With a composed ExecutionReadScope (C2 Stage 4 — the multi-tenant
 * edition), the scan narrows to the caller's authorized ids intersected
 * with the requested org (the Java GetExecutionSummary baseline: without
 * the org filter a member of several organizations sees every org's
 * numbers on every dashboard), and an EMPTY authorized set answers the
 * proto default instance — the conformance-pinned multi-tenant zero shape
 * (success_rate 0, NO cost summary) falls out of the scoping, never a
 * special case. No scope composed = the full scan above, byte-identical
 * (the request's org deliberately not consulted — single-user semantics).
 *
 * Tie order in the two ranked lists is not wire-stable in Go (map
 * iteration feeds a stable sort), so ties here — deterministic first-seen
 * order — are not a wire divergence (the #6 sort-stability precedent).
 */
import { create } from "@bufbuild/protobuf";
import { DurationSchema } from "@bufbuild/protobuf/wkt";
import type { Timestamp } from "@bufbuild/protobuf/wkt";
import { timestampMs } from "@bufbuild/protobuf/wkt";

import type { WorkflowExecution } from "@stigmer/protos/ai/stigmer/agentic/workflowexecution/v1/api_pb";
import { ExecutionPhase } from "@stigmer/protos/ai/stigmer/agentic/workflowexecution/v1/enum_pb";
import {
  ExecutionSummarySchema,
  SummaryTimeWindow,
  WorkflowCostBreakdownSchema,
  WorkflowCostSummarySchema,
  WorkflowFailureRankSchema,
} from "@stigmer/protos/ai/stigmer/agentic/workflowexecution/v1/io_pb";
import type {
  ExecutionSummary,
  GetExecutionSummaryRequest,
  WorkflowCostBreakdown,
  WorkflowFailureRank,
} from "@stigmer/protos/ai/stigmer/agentic/workflowexecution/v1/io_pb";

import { ApiResourceKind } from "@stigmer/protos/ai/stigmer/commons/apiresource/apiresourcekind/api_resource_kind_pb";

import type { Logger } from "../../boot/logger.js";
import type { ExecutionReadScope } from "../../extensions/execution-read-scope.js";
import type { CallerIdentity } from "../../extensions/identity.js";
import type { Store } from "../../store/interface.js";

import { parseRfc3339Ms } from "./execution-filter.js";
import { loadAllWorkflowExecutions } from "./queries.js";

const RANK_LIMIT = 10;

export interface SummaryDeps {
  readonly store: Store;
  readonly logger: Logger;
  /** The composed summary read scope — undefined = the OSS full scan (C2 Stage 4). */
  readonly executionReadScope: ExecutionReadScope | undefined;
}

export async function getExecutionSummary(
  deps: SummaryDeps,
  req: GetExecutionSummaryRequest,
  identity: CallerIdentity,
): Promise<ExecutionSummary> {
  let executions = await loadAllWorkflowExecutions(
    deps.store,
    deps.logger,
    "failed to list workflow executions for summary",
  );

  if (deps.executionReadScope !== undefined) {
    const authorizedIds = await deps.executionReadScope.authorizedExecutionIds(
      identity,
      ApiResourceKind.workflow_execution,
    );
    if (authorizedIds.size === 0) {
      // The Java baseline's empty arm: the default instance, before any
      // aggregation — success_rate 0 and NO cost summary by construction.
      return create(ExecutionSummarySchema);
    }
    executions = executions.filter(
      (execution) =>
        authorizedIds.has(execution.metadata?.id ?? "") &&
        execution.metadata?.org === req.org,
    );
  }

  const cutoffMs = resolveTimeCutoffMs(req.timeWindow);
  const workflowFilter = req.workflowId;

  const phaseCounts: Record<number, number> = {};
  let activeCount = 0;
  let totalCount = 0;
  let completedCount = 0;
  let failedCount = 0;
  const completedDurationsMs: number[] = [];
  const failureCounts = new Map<string, number>();
  const workflowNames = new Map<string, string>();
  const execByWorkflow = new Map<string, number>();
  const costByWorkflow = new Map<string, number>();

  let totalCostMicros = 0n;
  let totalInputTokens = 0n;
  let totalOutputTokens = 0n;

  for (const execution of executions) {
    const createdAtMs = auditCreatedAtMs(
      execution.status?.audit?.specAudit?.createdAt,
    );
    if (
      createdAtMs !== undefined &&
      cutoffMs !== undefined &&
      createdAtMs < cutoffMs
    ) {
      continue;
    }

    if (
      workflowFilter !== "" &&
      execution.spec?.workflowId !== workflowFilter &&
      execution.spec?.workflowInstanceId !== workflowFilter
    ) {
      continue;
    }

    const phase =
      execution.status?.phase ?? ExecutionPhase.EXECUTION_PHASE_UNSPECIFIED;
    phaseCounts[phase] = (phaseCounts[phase] ?? 0) + 1;
    totalCount++;

    if (isActivePhase(phase)) {
      activeCount++;
    }
    if (phase === ExecutionPhase.EXECUTION_COMPLETED) {
      completedCount++;
    }
    if (phase === ExecutionPhase.EXECUTION_FAILED) {
      failedCount++;
    }

    // Go's extractWorkflowSlug reads the EXECUTION's metadata.slug (the
    // per-workflow grouping key the dashboard renders).
    const slug = execution.metadata?.slug ?? "";
    if (slug !== "") {
      execByWorkflow.set(slug, (execByWorkflow.get(slug) ?? 0) + 1);
      const name = execution.metadata?.name ?? "";
      if (name !== "") {
        workflowNames.set(slug, name);
      }
    }

    const execCostMicros = execution.status?.totalCostMicros ?? 0n;
    totalCostMicros += execCostMicros;
    totalInputTokens += execution.status?.totalInputTokens ?? 0n;
    totalOutputTokens += execution.status?.totalOutputTokens ?? 0n;
    if (slug !== "" && execCostMicros > 0n) {
      costByWorkflow.set(
        slug,
        (costByWorkflow.get(slug) ?? 0) + Number(execCostMicros) / 1_000_000,
      );
    }

    if (phase === ExecutionPhase.EXECUTION_COMPLETED) {
      const durationMs = completionDurationMs(execution);
      if (durationMs > 0) {
        completedDurationsMs.push(durationMs);
      }
    }

    if (phase === ExecutionPhase.EXECUTION_FAILED && slug !== "") {
      failureCounts.set(slug, (failureCounts.get(slug) ?? 0) + 1);
    }
  }

  let successRate = -1;
  const terminal = completedCount + failedCount;
  if (terminal > 0) {
    successRate = completedCount / terminal;
  }

  const summary = create(ExecutionSummarySchema, {
    activeCount,
    phaseCounts,
    totalCost: create(WorkflowCostSummarySchema, {
      totalCostUsd: Number(totalCostMicros) / 1_000_000,
      totalInputTokens,
      totalOutputTokens,
    }),
    totalCount,
    successRate,
  });

  if (completedDurationsMs.length > 0) {
    let totalMs = 0;
    for (const durationMs of completedDurationsMs) {
      totalMs += durationMs;
    }
    summary.avgDuration = durationFromMs(
      totalMs / completedDurationsMs.length,
    );
  }

  summary.topFailingWorkflows = buildFailureRanks(
    failureCounts,
    workflowNames,
    RANK_LIMIT,
  );
  summary.costByWorkflow = buildCostBreakdown(
    costByWorkflow,
    execByWorkflow,
    workflowNames,
    RANK_LIMIT,
  );

  return summary;
}

/**
 * resolveTimeCutoff — LAST_7D is the default for unspecified/unknown
 * windows; ALL_TIME is the "no cutoff" sentinel (Go's zero time,
 * undefined here).
 */
export function resolveTimeCutoffMs(
  timeWindow: SummaryTimeWindow,
): number | undefined {
  const nowMs = Date.now();
  switch (timeWindow) {
    case SummaryTimeWindow.LAST_24H:
      return nowMs - 24 * 60 * 60 * 1000;
    case SummaryTimeWindow.LAST_7D:
      return nowMs - 7 * 24 * 60 * 60 * 1000;
    case SummaryTimeWindow.LAST_30D:
      return nowMs - 30 * 24 * 60 * 60 * 1000;
    case SummaryTimeWindow.ALL_TIME:
      return undefined;
    default:
      return nowMs - 7 * 24 * 60 * 60 * 1000;
  }
}

function isActivePhase(phase: ExecutionPhase): boolean {
  return (
    phase === ExecutionPhase.EXECUTION_PENDING ||
    phase === ExecutionPhase.EXECUTION_IN_PROGRESS ||
    phase === ExecutionPhase.EXECUTION_PAUSED
  );
}

/** Go auditCreatedAt: nil timestamp → zero time (undefined here). */
function auditCreatedAtMs(createdAt: Timestamp | undefined): number | undefined {
  if (createdAt === undefined) {
    return undefined;
  }
  return timestampMs(createdAt);
}

/**
 * Go completionDuration: 0 when either timestamp is missing or
 * unparseable — those executions are excluded from the average (the
 * `> 0` guard at the call site).
 */
function completionDurationMs(execution: WorkflowExecution): number {
  const started = parseRfc3339Ms(execution.status?.startedAt ?? "");
  const completed = parseRfc3339Ms(execution.status?.completedAt ?? "");
  if (Number.isNaN(started) || Number.isNaN(completed)) {
    return 0;
  }
  return completed - started;
}

function durationFromMs(ms: number) {
  const truncatedMs = Math.trunc(ms);
  const seconds = Math.trunc(truncatedMs / 1000);
  const nanos = Math.trunc((truncatedMs - seconds * 1000) * 1_000_000);
  return create(DurationSchema, { seconds: BigInt(seconds), nanos });
}

/** Failure ranks by count descending, capped (stable; ties first-seen). */
function buildFailureRanks(
  counts: Map<string, number>,
  names: Map<string, string>,
  limit: number,
): WorkflowFailureRank[] {
  const entries = [...counts.entries()].sort((a, b) => b[1] - a[1]);
  return entries.slice(0, limit).map(([slug, count]) =>
    create(WorkflowFailureRankSchema, {
      workflowSlug: slug,
      workflowName: names.get(slug) ?? "",
      failureCount: count,
    }),
  );
}

/** Cost breakdown by total cost descending, capped (ties first-seen). */
function buildCostBreakdown(
  costs: Map<string, number>,
  execCounts: Map<string, number>,
  names: Map<string, string>,
  limit: number,
): WorkflowCostBreakdown[] {
  const entries = [...costs.entries()].sort((a, b) => b[1] - a[1]);
  return entries.slice(0, limit).map(([slug, cost]) =>
    create(WorkflowCostBreakdownSchema, {
      workflowSlug: slug,
      workflowName: names.get(slug) ?? "",
      totalCostUsd: cost,
      executionCount: execCounts.get(slug) ?? 0,
    }),
  );
}
